src/main/java/org/czekalski/petapiintegration/config/AnimalPageCaseFactory.java

```diff
@@ -0,0 +1,117 @@
+                    linkTo(methodOn(AnimalController.class).getDogsFromCity(stateId, cityId, size, page)).withSelfRel(),
+                    linkTo(methodOn(AnimalController.class).getDogsFromCity(stateId, cityId, size, page + 1)).withRel("next"));
+        }
+    }
+
+    private class AfterAllPages implements PaginationCase {
+
+        @Override
+        public EntityModel<AnimalResourcesListDto> calculateCase(
+                AnimalResourcesListDto animalResourcesListDto,
+                String stateId,
+                String cityId,
+                int size,
+                int page) {
+            return new EntityModel<>(animalResourcesListDto,
+                    WebMvcLinkBuilder.linkTo(methodOn(AnimalController.class).getDogsFromCity(stateId, cityId, size, animalResourcesListDto.getPagination().getTotalPages())).withRel("prev"));
+        }
+
+    }
+
+
+    private class UnspecifiedPage implements PaginationCase {
+        @Override
+        public EntityModel<AnimalResourcesListDto> calculateCase(
+                AnimalResourcesListDto animalResourcesListDto,
+                String stateId,
+                String cityId,
+                int size,
+                int page) {
+            return new EntityModel<>(animalResourcesListDto,
+                    linkTo(methodOn(AnimalController.class).getDogsFromCity(stateId, cityId, size, page)).withSelfRel());
+        }
+    }
+
+    public EntityModel<AnimalResourcesListDto> toModel(
+            AnimalResourcesListDto animalResourcesListDto,
+            String stateId,
+            String cityId) {
+
+        int currentPage = animalResourcesListDto.getPagination().getPageable().getPageNumber();
+        int totalPages = animalResourcesListDto.getPagination().getTotalPages();
+        if (currentPage > 1 && currentPage == totalPages) {
+            return new LastPage().calculateCase(animalResourcesListDto, stateId, cityId, animalResourcesListDto.getPagination().getSize(), currentPage);
+        } else if (currentPage > 1 && currentPage > totalPages) {
+            return new AfterAllPages().calculateCase(animalResourcesListDto, stateId, cityId, animalResourcesListDto.getPagination().getSize(), currentPage);
+        } else if (currentPage > 1 && currentPage < totalPages) {
+            return new MiddlePage().calculateCase(animalResourcesListDto, stateId, cityId, animalResourcesListDto.getPagination().getSize(), currentPage);
+        } else if (currentPage == 1 && totalPages > 1) { //always exists at least 2 pages inside PageImpl
+            return new FirstPage().calculateCase(animalResourcesListDto, stateId, cityId, animalResourcesListDto.getPagination().getSize(), currentPage);
+        } else {
+            return new UnspecifiedPage().calculateCase(animalResourcesListDto, stateId, cityId, animalResourcesListDto.getPagination().getSize(), currentPage);
```

Comment:
Takie konstrukcje najlepiej jest rozwiązać fabryką. Fabryka generuje nam obiekt typu Page. Na podstawie tych warunków wybiera konkretną implementację. A na tym obiekcie typu Page wywołujemy metodą `calculateCase`.

Mogłoby to wyglądać tak
```
Page page = PageFactory.of(currentPage, totalPages)
return page.calculateCase(animalResourcesListDto, stateId, cityId, currentPage)
```
ten size można wyciągnąć wewnątrz calculateCase, bo lista jest przekazywana

Reply:
Poprawione: 0af844de692538af5d524ca9d17e6776ec2f0316